cpp/src/arrow/types/struct.cc

```diff
@@ -17,4 +17,94 @@
 
 #include "arrow/types/struct.h"
 
-namespace arrow {}  // namespace arrow
+#include <sstream>
+
+namespace arrow {
+
+bool StructArray::EqualsExact(const StructArray& other) const {
+  if (this == &other) { return true; }
+  if (null_count_ != other.null_count_) { return false; }
+
+  bool equal_null_bitmap = true;
+  if (null_count_ > 0) {
+    equal_null_bitmap =
+        null_bitmap_->Equals(*other.null_bitmap_, util::bytes_for_bits(length_));
+  }
+
+  if (!equal_null_bitmap) { return false; }
+  bool values_equal = false;
+  for (size_t i = 0; i < field_arrays_.size(); ++i) {
+    values_equal = field_arrays_.at(i).get()->Equals(other.fields().at(i));
```

Comment:
We need to expand the API to take an additional null bitmap that is needed to be logically anded with both arrays null bitmap.  I think this change will be fairly involved so likely best to handle as a separate JIRA and come back and fix this code after it is done.  I will open up JIRAs describing the problem.